File: hw/paillacho/U3/HW17Principles/MazeCode/Maze/src/main/java/ec/edu/espe/controller/DFSMazeGenerator.java
package ec.edu.espe.controller;

import ec.edu.espe.model.MazeModel;
import java.util.Random;
import java.util.Stack;
/**
 *
 * @author Carlos Paillacho
 */
public class DFSMazeGenerator implements MazeGenerator {
    @Override
    public MazeModel generate(int width, int height) {
        MazeModel maze = new MazeModel(width, height);
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                maze.setRoom(x, y, new Room(x, y));
            }
        }
        Stack<Room> stack = new Stack<>();
        Room start = maze.getRoom(0, 0);
        stack.push(start);
        Random rand = new Random();

        while (!stack.isEmpty()) {
            Room current = stack.peek();

            Room neighbor = /* ... */ null;
            if (neighbor != null) {

            } else {
                stack.pop();
            }
        }

        maze.setEntrance(new Entrance(maze.getRoom(0, 0)));
        maze.setExit(new Exit(maze.getRoom(width - 1, height - 1)));

        return maze;
    }
}

File: hw/paillacho/U3/HW17Principles/MazePython/controller/DFSMazeGenerator.py
from random import shuffle
from controller.Direction import directions


class DFSMazeGenerator:
    def generate(self, grid, width, height):
        visited = [[False for _ in range(height)] for _ in range(width)]
        self._dfs(grid, visited, 0, 0, width, height)

    def _dfs(self, grid, visited, x, y, width, height):
        visited[x][y] = True
        dirs = directions()
        shuffle(dirs)

        for dx, dy in dirs:
            nx, ny = x + dx, y + dy
            if 0 <= nx < width and 0 <= ny < height and not visited[nx][ny]:
                grid[x][y].remove_wall(dx, dy)
                grid[nx][ny].remove_wall(-dx, -dy)
                self._dfs(grid, visited, nx, ny, width, height)
